File: py/sort/array/StoogeSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class StoogeSort:
    def sort(self, nums: List[int]) -> None:
        self.__sort(nums, 0, len(nums) - 1)

    def __sort(self, nums: List[int], start: int, end: int) -> None:
        if nums[start] > nums[end]:
            nums[start], nums[end] = nums[end], nums[start]

        if start + 1 < end:
            third = (end - start + 1) // 3
            self.__sort(nums, start, end - third)
            self.__sort(nums, start + third, end)
            self.__sort(nums, start, end - third)

File: java/sort/array/StoogeSort.java
package sort.array;

public class StoogeSort {
    public void sort(int[] nums) {
        sort(nums, 0, nums.length - 1);
    }

    private void sort(int[] nums, int start, int end) {
        if (nums[start] > nums[end]) {
            nums[start] ^= nums[end];
            nums[end] ^= nums[start];
            nums[start] ^= nums[end];
        }

        if (start + 1 < end) {
            int third = (end - start + 1) / 3;
            sort(nums, start, end - third);
            sort(nums, start + third, end);
            sort(nums, start, end - third);
        }
    }
}
